beacon-chain/slasher/service.go

```diff
@@ -16,61 +18,91 @@ import (
 // parameters for slasher to function as needed.
 type ServiceConfig struct {
-	IndexedAttsFeed       *event.Feed
-	BeaconBlocksFeed      *event.Feed
-	AttesterSlashingsFeed *event.Feed
-	ProposerSlashingsFeed *event.Feed
-	Database              db.Database
-	GenesisTime           time.Time
+	IndexedAttestationsFeed *event.Feed
+	BeaconBlockHeadersFeed  *event.Feed
+	AttesterSlashingsFeed   *event.Feed
+	ProposerSlashingsFeed   *event.Feed
+	Database                db.Database
+	StateNotifier           statefeed.Notifier
 }
 
 // Service defining a slasher implementation as part of
 // the beacon node, able to detect eth2 slashable offenses.
 type Service struct {
-	params           *Parameters
-	serviceCfg       *ServiceConfig
-	indexedAttsChan  chan *ethpb.IndexedAttestation
-	beaconBlocksChan chan *ethpb.SignedBeaconBlockHeader
-	attsQueue        *attestationsQueue
-	blksQueue        *blocksQueue
-	ctx              context.Context
-	cancel           context.CancelFunc
-	genesisTime      time.Time
-	slotTicker       slotutil.Ticker
+	params                 *Parameters
+	serviceCfg             *ServiceConfig
+	indexedAttsChan        chan *ethpb.IndexedAttestation
+	beaconBlockHeadersChan chan *ethpb.SignedBeaconBlockHeader
+	attsQueue              *attestationsQueue
+	blksQueue              *blocksQueue
+	ctx                    context.Context
+	cancel                 context.CancelFunc
+	slotTicker             *slotutil.SlotTicker
 }
 
 // New instantiates a new slasher from configuration values.
 func New(ctx context.Context, srvCfg *ServiceConfig) (*Service, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	return &Service{
-		params:           DefaultParams(),
-		serviceCfg:       srvCfg,
-		indexedAttsChan:  make(chan *ethpb.IndexedAttestation, 1),
-		beaconBlocksChan: make(chan *ethpb.SignedBeaconBlockHeader, 1),
-		attsQueue:        newAttestationsQueue(),
-		blksQueue:        newBlocksQueue(),
-		ctx:              ctx,
-		cancel:           cancel,
-		genesisTime:      time.Now(),
+		params:                 DefaultParams(),
+		serviceCfg:             srvCfg,
+		indexedAttsChan:        make(chan *ethpb.IndexedAttestation, 1),
+		beaconBlockHeadersChan: make(chan *ethpb.SignedBeaconBlockHeader, 1),
+		attsQueue:              newAttestationsQueue(),
+		blksQueue:              newBlocksQueue(),
+		ctx:                    ctx,
+		cancel:                 cancel,
 	}, nil
 }
 
 // Start listening for received indexed attestations and blocks
 // and perform slashing detection on them.
 func (s *Service) Start() {
-	log.Info("Starting slasher")
-	s.slotTicker = slotutil.NewSlotTicker(s.genesisTime, params.BeaconConfig().SecondsPerSlot)
+	var genesisTime time.Time
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := s.serviceCfg.StateNotifier.StateFeed().Subscribe(stateChannel)
+	event := <-stateChannel
+
+	// Wait for us to receive the genesis time via a chain started notification.
+	if event.Type == statefeed.ChainStarted {
```

Comment:
Are these conditions covered in unit tests? If not, we'll be getting poor test coverage

Reply:
I can add some tests that check logs